### ISSUE TITLE
Handling navigation commands

### ISSUE THREAD

**Issue by noctuid** [polymode/polymode#165]
Regarding comment in #164.

> The most straightforward way to tackle that is to advise those function to operate within narrowed chunks.

I don't recall this being the case before, but maybe I never tried it. I wouldn't mind if the commands only moved between chunks, but it seems like that would be hard to do and would only make sense for some commands. 

I think this is a bigger issue about the best way to handle navigation commands. A lot of navigation commands should work across chunks, but not all do. For example, `worf-goto` and `counsel-org-goto` do not work in elisp blocks. I have a command in my init that calls `polymode-previous-chunk` if `worf-goto` fails. Maybe it would be nice if polymode provided two commands for advising navigation commands that did something like this:

1. Move outside of source block to a host chunk if necessary before running command (for commands like `counsel-org-goto`)
2. Prevent command from doing anything if it tries to move outside the current chunk (for lispy commands)

**Comment by vspinu** [comment]
There are a ton of navigation commands in various packages. There is no way polymode could deal with all those. It provides some compatibility layer with widely used packages and emacs built-ins, but in the long run individual packages should detect if they run in polymode buffers and deal with chunks themselves. 

There could be custom variables like `polymode-navigation-commands` to which users could add their commands and polymode will automatically advice them, but I haven't had time to think about this part of the API yet.

**Comment by vspinu** [comment]
> Maybe it would be nice if polymode provided two commands for advising navigation commands 

Indeed. This should be a decent start of the user level API. I have a bunch of advises in `polymode-compat.el` but nothing specific for navigation.  Regarding your  point 2, there is a common advice for narrowing to chunk before execution. This is how ess functions are currently advised:

```el
(advice-add 'ess-eval-paragraph :around 'pm-execute-narrowed-to-span)
(advice-add 'ess-eval-buffer :around 'pm-execute-narrowed-to-span)
(advice-add 'ess-beginning-of-function :around 'pm-execute-narrowed-to-span)
```

Could you try locally if it works for lispy and other functions for which it makes sense.

Regarding 1, I presume those functions have a check for major mode or something, otherwise they would work in all buffers.  There is currently an advice `#'polymode-with-current-base-buffer`. For example:

```el
(advice-add 'kill-buffer :around #'polymode-with-current-base-buffer)
```

As for 2, if you could identify all those functions in councel, worf etc which need such advises we could add it to the core in some custom `polymode-execute-in-host-buffer-functions` or something like that.

**Comment by vspinu** [comment]
Yeh. Ideally it should be done on the other package side. The question is how to do it in a cleaner and forward compatible way. 

I like the variable approach (probably not custom though) because both users and other pacakges can append to that variable and not bother with internals. Today the fix is advice-add, but maybe tomorrow the advice won't be needed. 

> A potential alternative approach would be to advise all commands for a package prefix. 

For packages like lispy or smart-parens this might make sense indeed. Maybe the custom variable could contain both symbols for actual functions or regexp to match the command. Also it's not entirely clear to me where to initialize the advices. If a package is loaded after polymode, the advice won't take place.

**Comment by vspinu** [comment]
BTW, the naming for these advices is a bit messy and will change.  Polylmode's user facing functions and hooks are named with `polymode-` prefix but development API with `pm-` prefix. ATM I cannot yet decide if those advises are user facing or part of the dev API.

**Comment by noctuid** [comment]
> Today the fix is advice-add, but maybe tomorrow the advice won't be needed.

As long as polymode provides functions (that actually call `advice-add` instead of the user or package) and/or variables to do whatever setup is necessary, other packages won't need to care about the internals.

> For packages like lispy or smart-parens this might make sense indeed. Maybe the custom variable could contain both symbols for actual functions or regexp to match the command. 

I'll try advising everything that lispy provides and see if I encounter any issues.

> Also it's not entirely clear to me where to initialize the advices. If a package is loaded after polymode, the advice won't take place.

Right, if that happens polymode will either need to know the package name, so it can add the advice in an `eval-after-load` or it will have to do something like use `after-load-functions`.

**Comment by vspinu** [comment]
BTW, I have just pushed major changes to poly-org and the core. The org host now runs in so called "unprotected" mode which should result in a much better experience.

**Comment by noctuid** [comment]
I didn't remember how to get interned symbols, but two people on \#emacs answered my question in about two seconds (`mapatoms` or `apropos-internal`). I'm testing this:
```elisp
(with-eval-after-load 'lispy
  (mapc (lambda (x)
            (advice-add x :around #'pm-execute-narrowed-to-span))
          (append (apropos-internal "^lispy-" #'commandp)
                  (apropos-internal "^special-lispy-" #'commandp))))
```

I'll report back how well this works (seems to work well in the basic cases).

**Comment by noctuid** [comment]
Restricting to a span seems to be working well. Executing with the current base buffer is no longer working for some commands. For example, if I run `outline-toggle-children` (advised with `polymode-with-current-base-buffer`) in an org source block nothing happens until the point leaves the source block. Then it closes the heading. I'm having other issues as well where commands like `counsel-outline` work inconsistently (usually it does nothing but the context is now the base buffer, e.g. keybindings for org mode are active instead of elisp).
